File: boj/TwoPointer/BOJ_1806.cpp
/*
 백준 1806 부분합
 21.07.07
 https://github.com/skyqnaqna/algorithm_study/
*/

#include <bits/stdc++.h>

#define INF 2147000000
#define endl "\n"

typedef long long ll;
typedef double dd;
typedef std::pair<int, int> pii;
typedef std::pair<ll, ll> pll;

using namespace std;

int n, s;
int arr[100001];

int main()
{
    ios::sync_with_stdio(0); cin.tie(0);

    cin >> n >> s;

    for (int i = 0; i < n; ++i) cin >> arr[i];

    int l = 0, r = 0, sum = 0, ans = INF;

    while (l < n) {
        if (r == n) {
            sum -= arr[l++];
        } else if (sum < s) sum += arr[r++];
        else sum -= arr[l++];

        if (sum >= s) ans = min(ans, r - l);
    }

    if (ans < INF) cout << ans << endl;
    else cout << 0 << endl;

    return 0;
}

File: boj/TwoPointer/BOJ_1806.swift
/*
 백준 1806 부분합
 21.12.17
 https://github.com/skyqnaqna/algorithm_study
 */

import Foundation

let input = readLine()!.split(separator: " ").map { Int(String($0))! }
let n = input[0], s = input[1]

let arr = readLine()!.split(separator: " ").map { Int(String($0))! }

var l = 0, r = 0, sum = 0, answer = arr.count + 1

while l < n {
  if r == n {
    sum -= arr[l]
    l += 1
  } else if sum < s {
    sum += arr[r]
    r += 1
  } else {
    sum -= arr[l]
    l += 1
  }
  
  if sum >= s && r - l < answer {
    answer =  r - l
  }
}

if answer == arr.count + 1 {
  print(0)
} else {
  print(answer)
}
